File: backend/app/services/scoring_service.py
from typing import Dict, List
import numpy as np

class ScoringService:
    def __init__(self):
        self.job_roles = {
            'software_engineer': ['python', 'java', 'sql', 'git', 'agile', 'javascript'],
            'data_scientist': ['python', 'sql', 'machine learning', 'statistics', 'pandas', 'tensorflow'],
            'devops_engineer': ['docker', 'kubernetes', 'aws', 'ci/cd', 'linux', 'terraform'],
            'frontend_dev': ['react', 'javascript', 'html', 'css', 'angular', 'vue.js']
        }
    
    def calculate_resume_score(self, skills: List[str], experience: str, education: str, 
                               ats_score: float, keyword_score: float) -> float:
        """Calculate overall resume score"""
        skill_score = self._calculate_skill_score(skills)
        exp_score = self._calculate_experience_score(experience)
        edu_score = self._calculate_education_score(education)
        
        # Weighted average
        total_score = (skill_score * 0.35 + exp_score * 0.30 + 
                      edu_score * 0.20 + ats_score * 0.15)
        
        return round(total_score, 2)
    
    def _calculate_skill_score(self, skills: List[str]) -> float:
        """Calculate skill relevance score"""
        if not skills:
            return 0
        # Assuming 10+ skills is perfect score
        base_score = min(len(skills) * 10, 100)
        return base_score
    
    def _calculate_experience_score(self, experience: str) -> float:
        """Calculate experience score"""
        if not experience:
            return 0
        
        # Extract years of experience
        import re
        years_pattern = r'(\d+)\+?\s*years?'
        matches = re.findall(years_pattern, experience, re.IGNORECASE)
        
        if matches:
            years = int(matches[0])
            return min(years * 10, 100)
        
        # If no years mentioned, give partial score
        return 50 if len(experience) > 50 else 25
    
    def _calculate_education_score(self, education: str) -> float:
        """Calculate education score"""
        if not education:
            return 0
        
        education_levels = {
            'phd': 100,
            'master': 90,
            'bachelor': 80,
            'diploma': 60,
            'certification': 50
        }
        
        edu_lower = education.lower()
        for level, score in education_levels.items():
            if level in edu_lower:
                return score
        
        return 40
    
    def generate_improvement_suggestions(self, resume_data: Dict) -> List[str]:
        """Generate personalized improvement suggestions"""
        suggestions = []
        
        # Skill-based suggestions
        if len(resume_data.get('skills', [])) < 5:
            suggestions.append("Add more relevant technical skills to improve your profile visibility")
        
        # Experience suggestions
        if not resume_data.get('experience'):
            suggestions.append("Include detailed work experience with specific achievements and metrics")
        
        # ATS suggestions
        if resume_data.get('ats_score', 0) < 60:
            suggestions.append("Optimize your resume for ATS by including standard section headings and relevant keywords")
        
        # Formatting suggestions
        if resume_data.get('formatting_score', 0) < 50:
            suggestions.append("Use bullet points and consistent formatting to improve readability")
        
        # Keyword suggestions
        if resume_data.get('keyword_score', 0) < 50:
            suggestions.append("Include more industry-specific keywords relevant to your target role")
        
        # Action verbs
        suggestions.append("Use strong action verbs to describe your accomplishments (e.g., 'achieved', 'implemented', 'developed')")
        
        # Quantifiable achievements
        suggestions.append("Include quantifiable achievements with numbers and percentages to demonstrate impact")
        
        return suggestions

File: backend/src/services/scoringService.js
class ScoringService {
  constructor() {
    this.jobRoles = {
      software_engineer: ['python', 'java', 'sql', 'git', 'agile', 'javascript'],
      data_scientist:    ['python', 'sql', 'machine learning', 'statistics', 'pandas', 'tensorflow'],
      devops_engineer:   ['docker', 'kubernetes', 'aws', 'ci/cd', 'linux', 'terraform'],
      frontend_dev:      ['react', 'javascript', 'html', 'css', 'angular', 'vue.js'],
    };
  }

  /**
   * Calculate weighted overall resume score (0-100).
   * @param {string[]} skills
   * @param {string}   experience
   * @param {string}   education
   * @param {number}   atsScore
   * @param {number}   keywordScore
   * @returns {number}
   */
  calculateResumeScore(skills, experience, education, atsScore, keywordScore) {
    const skillScore = this._calculateSkillScore(skills);
    const expScore   = this._calculateExperienceScore(experience);
    const eduScore   = this._calculateEducationScore(education);

    const total =
      skillScore * 0.35 +
      expScore   * 0.30 +
      eduScore   * 0.20 +
      atsScore   * 0.15;

    return Math.round(total * 100) / 100;
  }

  _calculateSkillScore(skills) {
    if (!skills || skills.length === 0) return 0;
    return Math.min(skills.length * 10, 100);
  }

  _calculateExperienceScore(experience) {
    if (!experience) return 0;

    const match = experience.match(/(\d+)\+?\s*years?/i);
    if (match) {
      return Math.min(parseInt(match[1], 10) * 10, 100);
    }

    return experience.length > 50 ? 50 : 25;
  }

  _calculateEducationScore(education) {
    if (!education) return 0;

    const levels = { phd: 100, master: 90, bachelor: 80, diploma: 60, certification: 50 };
    const lower = education.toLowerCase();
    for (const [level, score] of Object.entries(levels)) {
      if (lower.includes(level)) return score;
    }
    return 40;
  }

  /**
   * Generate personalised improvement suggestions.
   * @param {object} resumeData
   * @returns {string[]}
   */
  generateImprovementSuggestions(resumeData) {
    const suggestions = [];

    if ((resumeData.skills || []).length < 5) {
      suggestions.push('Add more relevant technical skills to improve your profile visibility');
    }

    if (!resumeData.experience) {
      suggestions.push('Include detailed work experience with specific achievements and metrics');
    }

    if ((resumeData.ats_score || 0) < 60) {
      suggestions.push('Optimize your resume for ATS by including standard section headings and relevant keywords');
    }

    if ((resumeData.formatting_score || 0) < 50) {
      suggestions.push('Use bullet points and consistent formatting to improve readability');
    }

    if ((resumeData.keyword_score || 0) < 50) {
      suggestions.push('Include more industry-specific keywords relevant to your target role');
    }

    suggestions.push('Use strong action verbs to describe your accomplishments (e.g., \'achieved\', \'implemented\', \'developed\')');
    suggestions.push('Include quantifiable achievements with numbers and percentages to demonstrate impact');

    return suggestions;
  }
}

module.exports = new ScoringService();
